### PR TITLE
Support manipulating JAR files with java.util.jar

### PR DESCRIPTION
**Pull request by Duhemm**
This commit adds support for almost all the classes defined in
`java.util.jar` (`JarInputStream` and `JarOutputStream` are missing and
will be added in a subsequent commit.)

Note that this implementation doesn't support verifying JARs signatures.
Everything is already supported on the `java.util.jar` side, but the
classes in `java.security` are only stubs. Implementing these classes
should be sufficient to verify JAR signatures.

This code has been ported from Apache Harmony.

### CODE REVIEW

File: javalib/src/main/scala/java/security/MessageDigest.scala [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package java.security
+
+// This is only a partial, minimal implementation.
+// TODO: Port me from Apache Harmony
```
Comment (Duhemm) [inline javalib/src/main/scala/java/security/MessageDigest.scala]: I added a note to the issue tracking progress on the zip/jar front (#932) to remind us to eventually port those for real.

File: javalib/src/main/scala/java/util/jar/JarVerifier.scala [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+    (metaEntries.get(signatureFile), metaEntries.get(JarFile.MANIFEST_NAME)) match {
+      case (null, _) | (_, null) =>
+        ()
+      case (sfBytes, manifest) =>
+        val sBlockBytes = metaEntries.get(certFile)
+        try {
+          // TODO: Port JarUtils from Apache Harmony.
+          // val signerCertChain = JarUtils.verifySignature(
+          //   new ByteArrayInputStream(sfBytes),
+          //   new ByteArrayInputStream(sBlockBytes))
```
Comment (Duhemm) [inline javalib/src/main/scala/java/util/jar/JarVerifier.scala]: This is mentioned in #932.

File: unit-tests/src/test/scala/java/util/jar/JarEntrySuite.scala [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    val attrJar       = getAttJarFile()
+    val attrsJarEntry = attrJar.getJarEntry(attEntryName)
+    assert(attrsJarEntry.getAttributes() != null)
+
+    val noAttrsJarEntry = attrJar.getJarEntry(attEntryName2)
+    assert(noAttrsJarEntry.getAttributes() == null)
+    attrJar.close()
+  }
+
+  // TODO: Uncomment once we support signed jars
```
Comment (Duhemm) [inline unit-tests/src/test/scala/java/util/jar/JarEntrySuite.scala]: Mentioned in #932

File: unit-tests/src/test/scala/java/util/jar/JarFileSuite.scala [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+    enumeration = jarFile.entries()
+    assert(enumeration.hasMoreElements())
+    val je = enumeration.nextElement()
+    jarFile.close()
+    assertThrows[IllegalStateException] {
+      jarFile.getJarEntry(je.getName)
+    }
+  }
+
+  // TODO: Enable this test once we support `JarOutputStream`:
```
Comment (Duhemm) [inline unit-tests/src/test/scala/java/util/jar/JarFileSuite.scala]: I've already amended my port of `JarOutputStream` to re-enable this test.

File: unit-tests/src/test/scala/java/util/jar/JarEntrySuite.scala (hunk 3 above)
Comment (densh) [inline unit-tests/src/test/scala/java/util/jar/JarEntrySuite.scala]: We have `testFails("...", issue = 12345) { ... }` to predicate tests that are expected to fail with link to an issue. Lets use that instead for all of the TODO tests.

**Comment by densh** [comment]
Please rebase your PR on top of latest master due to CI changes, thanks!

**Comment by Duhemm** [comment]
@densh Rebased & opened issues.

**Comment by densh** [comment]
LGTM, thanks @Duhemm ! 🍰

**Pull request merged**